### PR TITLE
Revamp

### PR DESCRIPTION
**Pull request by ehsanh06**
I've rewritten this again - opened up a branch as revamp. Trouble I'm facing is getting the MD's to generate at the gulp post -t "title" or gulp page -t "title", despite there being no errors. 

Not sure why my views folder hasn't been pushed

### CODE REVIEW

**Comment by ehsanh06** [comment]
Believe to have found what's causing it --seems to be the directory that's pointing towards that's not working. I'll have to work out exactly what directory it needs to be pointed to. Reason being is because there's a lot of requires for the global js file, and I may need to move that. Would like to know what you think about it.

File: gulpfile.js [hunk 1]
```diff
@@ -1,87 +1,18 @@
...
-// });
-
-
-// Sass gulp task
-gulp.task('sass-build', function () {
-    return gulp.src(files.input)
-        .pipe(sass({outputStyle: 'compressed'}))
-        .pipe(rename(files.output))
-        .pipe(gulp.dest(paths.distCss));
+// Include Gulp and modular task file
```
Comment (shahidaltaf) [inline gulpfile.js]: Add 'use strict';

File: gulpfile.js [hunk 2]
```diff
@@ -1,87 +1,18 @@
...
+var gulp = require('gulp');
+var tasks = require('./lib/tasks/index');
```
Comment (shahidaltaf) [inline gulpfile.js]: This can be simply `var tasks = require('./lib/tasks/');`

File: gulpfile.js [hunk 3]
```diff
@@ -1,87 +1,18 @@
...
+
+// Gulp tasks
+gulp.task('post', function() {
+    console.log('Console Message: Building Post'),
```
Comment (shahidaltaf) [inline gulpfile.js]: The console log here is redundant. it would be better placed in the `tasks.post`

File: gulpfile.js [hunk 4]
```diff
@@ -1,87 +1,18 @@
...
+    tasks.post;
```
Comment (shahidaltaf) [inline gulpfile.js]: `tasks.post;` doesn't call the function, it should be `tasks.post();`

Comment (shahidaltaf) [reply gulpfile.js]: `tasks.post();` should take a **title** parameter, and an optional **slug**

File: gulpfile.js [hunk 5]
```diff
@@ -1,87 +1,18 @@
...
 });
-
-gulp.task('clean', function() { rmrf.sync(paths.dist); });
-
-gulp.task('build', ['clean', 'site-build', 'sass-build'], function() {
-   console.log('Task success: All pages built successfully.');
+gulp.task('page', function() {
+    console.log('Console Message: Building Page'),
+    tasks.page;
```
Comment (shahidaltaf) [inline gulpfile.js]: `tasks.page;` doesn't call the function, it should be `tasks.page();`

Comment (shahidaltaf) [reply gulpfile.js]: Should take a title parameter

File: gulpfile.js [hunk 6]
```diff
@@ -1,87 +1,18 @@
...
-
-gulp.task('watch:sass', function() { gulp.watch(paths.sass, ['sass-build']); });
-
-
-// Default task
-gulp.task('default', ['build', 'watch:sass'], function() { console.log('Default tasks'); });
\ No newline at end of file
+gulp.task('build', function() {
+    console.log('Console Message: Building in progress'),
+    tasks.build;
```
Comment (shahidaltaf) [inline gulpfile.js]: `tasks.build;` doesn't call the function, it should be `tasks.build();`

File: lib/global.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
+// Include Gulp, plugins and local files
```
Comment (shahidaltaf) [inline lib/global.js]: `global.js` should be renamed as `utils.js` and placed within the `tasks` folder

**Review by shahidaltaf — CHANGES_REQUESTED** [review]

File: gulpfile.js [hunk 8]
```diff
@@ -1,87 +1,20 @@
...
-
-// Sass gulp task
-gulp.task('sass-build', function () {
-    return gulp.src(files.input)
-        .pipe(sass({outputStyle: 'compressed'}))
-        .pipe(rename(files.output))
-        .pipe(gulp.dest(paths.distCss));
+    // With default parameters in ES2015, the check in the function body is no longer necessary
+    // Slug is an optional parameter
+    tasks.post(title, slug);
```
Comment (shahidaltaf) [inline gulpfile.js:12]: `title` and `slug` are not declare variables in the scope of this file. Please catch command line arguments to pass into the `tasks.post(title, slug)` function

File: gulpfile.js [hunk 9]
```diff
@@ -1,87 +1,20 @@
...
 });
-
-gulp.task('clean', function() { rmrf.sync(paths.dist); });
-
-gulp.task('build', ['clean', 'site-build', 'sass-build'], function() {
-   console.log('Task success: All pages built successfully.');
+gulp.task('page', function() {
+    tasks.page(title);
```
Comment (shahidaltaf) [inline gulpfile.js:17]: `title` is not A declared variable in the scope of this file. Please catch command line argument to pass into the tasks.page(title) function

File: lib/tasks/index.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+var post = require('./post/post');
```
Comment (shahidaltaf) [inline lib/tasks/index.js]: `require('./post/post')...` are all invalid paths now as you have taken out sub folders

File: lib/tasks/index.js [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+var page = require('./page/page');
+var build = require('./build/build');
+
+module.exports = {
+    post: post,
+    page: page,
+    build: build
+};
```
Comment (shahidaltaf) [inline lib/tasks/index.js]: module exports can be (preferably) written as...

```
module.exports.post = require('./post');
module.exports.page = require('./page');
module.exports.build = require('./build');
```

File: lib/tasks/page.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+// Include Gulp, plugins and local files
+var argv = require('yargs').argv;
+var fs = require('fs');
+var gulp = require('gulp');
+var mkdirp = require('mkdirp');
+var moment = require('moment');
+
+var utils = require('../utils');
+
+function page() {
```
Comment (shahidaltaf) [inline lib/tasks/page.js]: `function page()` needs to take a `title` argument

File: lib/tasks/page.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    } else {
+
+        // Otherwise we log the error out to console if user didn't input -t or --title
+        return 'Error: You must enter a title for your page. Use --title "Page title" or -t "Page title"';
+    }
+
+}
+
+
+module.exports = {
```
Comment (shahidaltaf) [inline lib/tasks/page.js]: This can be better written as `module.exports = page` as there's only one function to export

File: lib/tasks/post.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        });
+
+    } else {
+
+        // Otherwise we log the error out to console if user didn't input -t or --title
+        return 'Error: You must enter a title for your page. Use --title "Page title" or -t "Page title"';
+    }
+};
+
+module.exports = {
```
Comment (shahidaltaf) [inline lib/tasks/post.js]: This can be better written as module.exports = post as there's only one function to export

File: lib/tasks/post.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+// Include Gulp, plugins and local files
+var argv = require('yargs').argv;
+var fs = require('fs');
+var gulp = require('gulp');
+var mkdirp = require('mkdirp');
+var moment = require('moment');
+
+var utils = require('../utils');
+
+function post() {
```
Comment (shahidaltaf) [inline lib/tasks/post.js]: function post() needs to take `title` and  optional `slug` arguments. These need to passed into the function and the handling of `argv.title` below needs to be revisted

File: lib/tasks/utils.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    drafts: '../../views/drafts/',
+    layouts: '../../views/layouts/',
+    includes: '../../public/src/js/',
+    sass: '../../public/src/sass/*.scss',
+    dist: '../../dist/',
+    distCss: '../../dist/css/'
+};
+
+// siteData will hold the meta information for the site
+var siteData = {
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: This data object is very specific to the build process and should be in `build.js`

File: lib/tasks/utils.js [hunk 17]
```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+// Include Gulp, plugins and local files
+var fs = require('fs');
+var handlebars = require('handlebars');
+var mkdirp = require('mkdirp');
+
+// Define Base folders 
+var srcDir = {
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: This `srcDir` object is very specific to the build process and should be in `build.js`

File: lib/tasks/utils.js [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    tags: [],
+    navigation: []
+};
+
+// Functions
+var build = function () {
+    console.log('Build Initiated');
+};
+
+var createPage = function (data) {
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: This is not a utility function and is very specific to the build process and should be in `build.js`

File: lib/tasks/utils.js [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+var siteData = {
+    posts: [],
+    pages: [],
+    archive: [],
+    tags: [],
+    navigation: []
+};
+
+// Functions
+var build = function () {
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: Remove this function, it is redundant

File: lib/tasks/utils.js [hunk 20]
```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+// Include Gulp, plugins and local files
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: Remove this comment

File: lib/tasks/utils.js [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+var fs = require('fs');
+var handlebars = require('handlebars');
+var mkdirp = require('mkdirp');
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: Remove unused modules

File: lib/tasks/utils.js [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+// Function which takes care of slug
+var hyphenateString = function (slug) {
+
+    // Convert slug to lowercase, trim whitespace, replace any /'s with '-'s
+    return slug.toLowerCase().trim().replace(/[^\w\s]/g, '').replace(/\s+/g, '-');
+};
+
+// Setting the page title
+var setPageTitle = function (title) { return title + ' - ' + srcDir.author; };
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: This function is specific to `build.js`, should this be here?

File: lib/tasks/utils.js [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+
+
+// Module Export methods
+module.exports = {
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: Update the exports module to only have the refactored utility functions in

**Review by shahidaltaf — CHANGES_REQUESTED** [review]
Just a structural review and request changes. I specific code review will follow

File: gulpfile.js [hunk 24]
```diff
@@ -1,87 +1,22 @@
...
-var rmrf        = require('rimraf');
-var sass        = require('gulp-sass');
+'use strict';
 
-var conf        = require('./conf.json');
-var dir         = requireDir('./tasks/');
+// Include Gulp and modular task file
+var gulp = require('gulp');
+var tasks = require('./lib/tasks/');
+var t = title || null,
```
Comment (shahidaltaf) [inline gulpfile.js]: This variables make no sense being here. This should be in the context of the gulp tasks below

Comment (shahidaltaf) [inline gulpfile.js]: Where is `title` declared? You've used `title` but you haven't declared it.

File: gulpfile.js [hunk 25]
```diff
@@ -1,87 +1,22 @@
...
-
-// Sass gulp task
-gulp.task('sass-build', function () {
-    return gulp.src(files.input)
-        .pipe(sass({outputStyle: 'compressed'}))
-        .pipe(rename(files.output))
-        .pipe(gulp.dest(paths.distCss));
+    // With default parameters in ES2015, the check in the function body is no longer necessary
+    // Slug is an optional parameter
+    tasks.post(title, slug);
```
Comment (shahidaltaf) [inline gulpfile.js:12]: `title` and `slug` should be defined above this method. It's ok to hard-code some default values here for now e.g. `var title = "Post example title"`. Same with slug.

File: lib/tasks/build.js [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    readMarkdownFiles();
+    createAuxData();
+    createAllPages();
+
+};
+
+module.exports = {
+    srcDir: srcDir,
```
Comment (shahidaltaf) [inline lib/tasks/build.js]: `srcDir` isn't a good enough object name, use full words

Comment (shahidaltaf) [inline lib/tasks/build.js]: `srcDir` and `siteData` have no need to be exported. Please remove them.

File: lib/tasks/index.js [hunk 27]
```diff
@@ -0,0 +1,7 @@
+var post = require('./post');
```
Comment (shahidaltaf) [inline lib/tasks/index.js]: Remove these `var` declarations. Now redundant and unused given the changes below.

File: lib/tasks/post.js [hunk 28]
```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+var argv = require('yargs').argv;
+var fs = require('fs');
+var mkdirp = require('mkdirp');
+var moment = require('moment');
+
+var utils = require('./utils');
+var build = require('./build');
```
Comment (shahidaltaf) [inline lib/tasks/post.js]: `build.js` isn'n need in this task file

File: lib/tasks/post.js [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+
+var t = title || null,
```
Comment (shahidaltaf) [inline lib/tasks/post.js]: Remove these declarations here as they should come from the function call below

File: lib/tasks/utils.js [hunk 30]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+
+// Checking if directory exists
+var dirExists = function (path, err) {
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: For utility functions, use full words and not abbreviations

File: lib/tasks/utils.js [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+    try {
+        fs.statSync(path);
```
Comment (shahidaltaf) [inline lib/tasks/utils.js]: `fs` module isn't declared above. You must add it to use `fs` functions

File: gulpfile.js [hunk 32]
```diff
@@ -1,87 +1,22 @@
...
-var sass        = require('gulp-sass');
+'use strict';
 
-var conf        = require('./conf.json');
-var dir         = requireDir('./tasks/');
+// Include Gulp and modular task file
+var gulp = require('gulp');
+var tasks = require('./lib/tasks/');
+var t = title || null, 
+    s = slug ||  null;
```
Comment (shahidaltaf) [inline gulpfile.js]: Same as comment for `title`, please remove and put `slug in the appropriate gulp task

File: gulpfile.js [hunk 33]
```diff
@@ -1,87 +1,22 @@
...
-gulp.task('watch:sass', function() { gulp.watch(paths.sass, ['sass-build']); });
-
-
-// Default task
-gulp.task('default', ['build', 'watch:sass'], function() { console.log('Default tasks'); });
\ No newline at end of file
+gulp.task('build', function() {
+    tasks.build();
+});
+// Default task goes here
```
Comment (shahidaltaf) [inline gulpfile.js]: Add a default task that simply console log out 'default task'

**Review by shahidaltaf — CHANGES_REQUESTED** [review]
Add a README.md file back in

File: lib/page.js [hunk 34]
```diff
@@ -5,32 +5,19 @@ var fs = require('fs');
...
+var title, slug = "Page title example";
 
 function page(title) {
     console.log('Console Message: Building Page');
-
-    // If the output is either -t or --title  then we store this to title
     if (title) {
-
-        var title = argv.title || argv.t;
-        var slug = (argv.slug || argv.s) ? utils.hyphenateString(argv.slug || argv.s) : utils.hyphenateString(title.toString());
```
Comment (shahidaltaf) [inline lib/page.js]: Restore this line please, you need to format a slug

File: lib/page.js [hunk 35]
```diff
@@ -5,32 +5,19 @@ var fs = require('fs');
 var mkdirp = require('mkdirp');
 var moment = require('moment');
 
+var config = require('./config.json');
 var utils = require('./utils');
-var build = require('./build');
 
-var t = title || null,
-    s = slug || null;
+var title, slug = "Page title example";
```
Comment (shahidaltaf) [inline lib/page.js]: Delete this as it's not required

File: lib/post.js [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+
+var argv = require('yargs').argv;
+var fs = require('fs');
+var mkdirp = require('mkdirp');
+var moment = require('moment');
+
+var config = require('./config.json');
+var utils = require('./utils');
+
+var title, slug = "Post title example";
```
Comment (shahidaltaf) [inline lib/post.js]: Delete this line as it's not required

File: gulpfile.js [hunk 37]
```diff
@@ -1,87 +1,25 @@
...
-var sass        = require('gulp-sass');
+'use strict';
 
-var conf        = require('./conf.json');
-var dir         = requireDir('./tasks/');
+var gulp = require('gulp');
+var tasks = require('./lib/tasks/');
 
+gulp.task('post', function() {
+    var title, slug = "Post title example";
```
Comment (shahidaltaf) [inline gulpfile.js]: Not sure why you have set `slug` to the same as the `title`. You either leave it out or set it to something specific . Keep the variables on separate lines for now

**Review by shahidaltaf — CHANGES_REQUESTED** [review]
A few minor oversights need addressing

**Pull request merged**